Top Tabs for iPad

tylerlacroix opened the pull request:
- Moved tabs button to Top Tabs bar
- Added new tab button
- Added private mode button
- Added Curve View for tabs

tylerlacroix commented:
Some screenshots of top tabs in action! 
<img width="880" alt="screen shot 2016-06-24 at 1 23 54 pm" src="https://cloud.githubusercontent.com/assets/11801330/16345330/6fcdc3a0-3a0f-11e6-8f08-2ba908db9032.png">
<img width="1136" alt="screen shot 2016-06-24 at 1 26 57 pm" src="https://cloud.githubusercontent.com/assets/11801330/16345334/72fbdd8c-3a0f-11e6-9508-ac92a96a673f.png">
<img width="1136" alt="screen shot 2016-06-24 at 1 27 14 pm" src="https://cloud.githubusercontent.com/assets/11801330/16345338/77b1eda8-3a0f-11e6-9baa-c3e5c19ee315.png">

sleroux commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -3168,6 +3241,7 @@ extension BrowserViewController: FindInPageBarDelegate, FindInPageHelperDelegate
     func findInPage(findInPage: FindInPageBar, didTextChange text: String) {
         find(text, function: "find")
     }
+    
```
Nit: Whitespace

sleroux commented on Client/Frontend/Browser/FaviconManager.swift:
```diff
@@ -80,6 +82,7 @@ class FaviconManager : TabHelper {
                         })
                     }
                 }
+            NSNotificationCenter.defaultCenter().postNotificationName(FaviconManager.FaviconDidLoad, object: nil)
```
You might experience a race condition with this notification because technically the favicon hasn't been added to the database yet when this gets fired. On https://github.com/mozilla/firefox-ios/pull/1946/files#diff-1dccca00bfd714476c188b2b63c6e659R76, a `Deferred` is return which wraps the async write to the database. If you want to make sure that the favicon is in the DB before firing you'll need to add an `upon` or `>>> effect()` block to it.

tylerlacroix replied:
Oh true! I didn't think of that, I'll fix it :)

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+
+class TopTabsViewController: UIViewController {
+    let tabManager: TabManager
+    weak var delegate: TopTabsDelegate!
```
Watch out for explicitly unwrapping `weak` variables.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+        privateTab.setImage(UIImage(named: "menu-NewPrivateTab-pbm"), forState: .Highlighted)
+        privateTab.accessibilityLabel = NSLocalizedString("New Tab", comment: "Accessibility label for the Private Tab button in the tab toolbar.")
+        return privateTab
+    }()
```
Nit: For these properties, I feel like these were copied from elsewhere to replicate the style we were using in the Tabs Tray. If the code identical, I would refactor this into an `extension` class method on UIButton/TabsButton to DRY up the code so we don't have it in multiple places.

tylerlacroix replied:
That makes sense! Where should I add the extension? TabsTray, TopTabs, or in a new file?

sleroux replied:
Probably a new file would make sense

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+    }
+    
+    override func viewWillAppear(animated: Bool) {
+        collectionView.dataSource = self
```
Make sure to call `super.viewWillAppear(animated)`

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+        self.tabsButton.updateTabCount(count, animated: animated)
+    }
+    
+    func tabsClicked() {
```
Nit: Rename to `tabsTapped` since we don't click on iPads 😛

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+    }
+    
+    override func viewDidAppear(animated: Bool) {
+        collectionView.reloadData()
```
Nit: For VC lifecycle events, call super prior to the subclass logic.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+        super.viewDidAppear(animated)
+    }
+    
+    func newTabClicked() {
```
Nit: Clicked -> Tapped

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+protocol TopTabsDelegate: class {
+    func topTabsPressTabs()
+    func topTabsPressNewTab()
+    func topTabsPressPrivateTab()
```
Nit: Minor thing but the Cocoa APIs usually include a more precise timing of the events such as Will/Did to the method to clarify exactly when we are interacting with an element. I would just name these to `topTabsDidPressNewTab` if we called it after pressing the tab button.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+        })
+    }
+    
+    func privateTabClicked() {
```
Nit: Clicked -> Tapped

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+        guard let currentTab = tabManager.selectedTab, let index = tabsToDisplay.indexOf(currentTab) else {
+            return
+        }
+        guard !collectionView.frame.isEmpty else {
```
Nit: You can merge this boolean check with the above guard by adding a `where` clause:

```swift
guard let currentTab = tabManager.selectedTab, let index = tabsToDisplay.indexOf(currentTab) where !collectionView.frame.isEmpty else {
    return
}
```

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+extension TopTabsViewController: Themeable {
+    func applyTheme(themeName: String) {
+        tabsButton.applyTheme(themeName)
+        if themeName == Theme.PrivateMode {
```
Nit: isPrivate = (themeNamed == Theme.PrivateMode)

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+
+extension TopTabsViewController: TopTabCellDelegate {
+    func tabCellDidClose(cell: TopTabCell) {
+        let indexPath = collectionView.indexPathForCell(cell)!
```
Nit: Probably not a problem but safer to wrap in a guard in case.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+        tabCell.titleText.text = tab.displayTitle
+        
+        if tab.displayTitle.isEmpty {
+            if (tab.url?.baseDomain()?.contains("localhost") ?? false || tab.url == nil) {
```
Nit: Don't think the nil check is necessary at the end since the optional will fail in the first expression and result in false.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+            }
+        }
+        
+        tabCell.selectedTab = tab == tabManager.selectedTab
```
Nit: I would just wrap the right hand expression in brackets just to add a bit of clarity to the statement.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+            tabCell.seperatorLine = false
+        }
+        
+        if !tab.displayTitle.isEmpty {
```
Nit: You already have an `if` statement above checking the inverse of this boolean. Maybe combine the two?

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+        }
+        
+        if let favIcon = tab.displayFavicon {
+            tabCell.favicon.sd_setImageWithURL(NSURL(string: favIcon.url)!)
```
Nit: I wonder if there is a case where the favicon URL passed in doesn't result in a proper URL? Best to add this to the optional unwrapping in the if statement.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+        self.favicon.layer.cornerRadius = 2.0
+        self.favicon.layer.masksToBounds = true
+        
+        self.titleText = UILabel()
```
Nit: Why not move this into a property closure like you did with `BezierView`?

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+        self.titleText.numberOfLines = 1
+        self.titleText.font = DynamicFontHelper.defaultHelper.DefaultSmallFontBold
+        
+        self.closeButton = UIButton()
```
Nit: Same comment as above.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+    }
+    
+    override func prepareForReuse() {
+        self.titleText.font = DynamicFontHelper.defaultHelper.DefaultSmallFontBold
```
Nit: Probably should call super here.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+private class TopTabsLayoutDelegate: NSObject, UICollectionViewDelegateFlowLayout {
+    weak var tabSelectionDelegate: TabSelectionDelegate?
+    
+    override init() {
```
Nit: Don't need this init if we're not doing anything in it.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+        collectionView.reloadData()
+    }
+    
+    func userContentController(userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
```
Does this get called anywhere? Just trying to figure out the motivation for implementing this delegate.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+        return attributes
+    }
+    
+    override func layoutAttributesForDecorationViewOfKind(elementKind: String, atIndexPath indexPath: NSIndexPath) -> UICollectionViewLayoutAttributes? {
```
Nit: Don't need to override if we don't do anything in it.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+class TopTabsBackgroundDecorationView : UICollectionReusableView {
+    static let Identifier = "TopTabsBackgroundDecorationViewIdentifier"
+    private lazy var right = SingleCurveView(right: true)
+    private lazy var left = SingleCurveView(right: false)
```
Nit: I would add `Curve` to these variables just to add some more clarity like `rightCurve` and `leftCurve`.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func layoutSubviews() {
```
Nit: Instead of overriding `layoutSubview` to call `setNeedsDisplay`, you can set the `contentMode` property in the init method to `.Redraw` which will do this for you.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+    
+    override func drawRect(rect: CGRect) {
+        super.drawRect(rect)
+    }
```
Actually I think using `contentMode = .Redraw` will eliminate all of these methods.

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,714 @@
+            let x2: CGFloat = 5.1
+            let x3: CGFloat = 19.76
+            let x4: CGFloat = 58.27
+            let x5: CGFloat = -12.15
```
This looks like the same code and points for the `BezierView`. Is there a way we can DRY this up instead of having it in multiple places?

sleroux commented:
Overall this PR looks great From a broader code perspective, I would look into splitting up `TopTabsViewController` into smaller, more specific files. Right now, the cell subclasses and views used inside of them are located in the middle of the file between the extension definitions. It might be a bit unwieldy to create a new .swift file for each cell but maybe a single .swift file containing cell/reusable views along with the views they make use of (BezierView, etc). Moving these internal helper views into private scope would also clean up the API a bit as well.

As for the collection view data source/delegates for `TopSitesViewController`, I think it's OK for them to extend directly off the view controller since the file isn't large after extracting the cell/view subclasses but in the future I would look into breaking these apart from the VC to avoid MVC (Massive View Controller).

sleroux commented on Client/Frontend/Browser/FaviconManager.swift:
```diff
@@ -31,14 +36,81 @@ class FaviconManager : TabHelper {
         return "faviconsMessageHandler"
     }
+    
+    internal class FaviconError: MaybeErrorType {
```
Nit: I would move this outside of this class scope. We don't really define our error types inside classes like this.

sleroux commented on Client/Frontend/Browser/FaviconManager.swift:
```diff
@@ -31,14 +36,81 @@ class FaviconManager : TabHelper {
+        
+        if let currentURL = tab.url,
+            let url = tab.url?.absoluteString {
+            let site = Site(url: url, title: "")
```
If this if statement fails and we return the deferred, the deferred will never finish since it will never be fulfilled. I would perform this if check outside the scope of this method and pass in the unwrapped values as parameters so you don't have to worry about checking this here.

sleroux commented on Client/Frontend/Browser/FaviconManager.swift:
```diff
@@ -31,14 +36,81 @@ class FaviconManager : TabHelper {
+                                                self.profile.favicons.addFavicon(fav, forSite: site).upon { _ in
+                                                    deferred.fill(Maybe(success: fav))
+                                                }
+                                                if tab.favicons.isEmpty {
```
This check will always fail because we append the favicon to `tab.favicons` a few lines up. I would double check with @jhugman about when/how we're supposed to use the `makeFaviconAvailable` method since it relates to the Spotlight extension.

sleroux replied:
I'm also a bit uneasy calling an async method then running this method afterwards which might depend on each other. calling `profile.favicons.addFavicon` should probably be the last thing we do.

tylerlacroix replied:
My bad!  I moved the "tab.favicons.append(fav)" up because I thought we were going to add it anyways, but I didn't realize we check if it's empty.  I'll move it back down and make the async call the last method.

sleroux commented on Client/Frontend/Browser/FaviconManager.swift:
```diff
@@ -31,14 +36,81 @@ class FaviconManager : TabHelper {
+                                                date: NSDate(),
+                                                type: icon.type)
+                                            
+                                            if let img = img {
```
Nit: I would replace this with a `guard` and move it to the first line of the closure to fail fast if we don't have an image

```swift
guard let img = img else {
    // fail
}
// Setup favicon
```

tylerlacroix replied:
Thank you so much for helping me with this deferred stuff! It can be quite confusing at times.

sleroux commented on Client/Frontend/Browser/TabButtons.swift:
```diff
@@ -0,0 +1,38 @@
+//
+//  Created by Tyler Lacroix on 6/27/16.
+//  Copyright © 2016 Mozilla. All rights reserved.
+//
```
MPL 2.0

sleroux replied:
Also I think the name of this file is a bit confusing since we also have `TabsButton`. Maybe rename this to TabTrayButtonExtensions and move to the Widgets folder.

sleroux commented on Client/Frontend/Browser/TopTabCell.swift:
```diff
@@ -0,0 +1,343 @@
+//  Created by Tyler Lacroix on 6/27/16.
+//  Copyright © 2016 Mozilla. All rights reserved.
+//
+
```
MPL 2.0

sleroux replied:
The file name doesn't really match whats going on in here. Might make more sense to name this 'TopTabViews' and extract the collection view delegate/datasource into their own file or files.

tylerlacroix replied:
I renamed it and separated the top tabs layout into a new file, does that make more sense or should I split up the VC file further? :)

sleroux commented on Client/Frontend/Browser/TopTabCell.swift:
```diff
@@ -0,0 +1,343 @@
+    }
+}
+
+class BezierView: UIView {
```
If this isn't used outside of this file make private

sleroux commented on Client/Frontend/Browser/TopTabsViewController.swift:
```diff
@@ -0,0 +1,340 @@
+        case Both
+    }
+    
+    static func TopTabsCurve(width: CGFloat, height: CGFloat, direction: TopTabsCurveDirection) -> UIBezierPath {
```
Nit: Now that we have a separate file for the views, I would move this and the enum to that file instead of the VC. Also, why not put this as an extension on `UIBezierPath`?

sleroux commented:
@tylerlacroix Almost there! Just some more nits and favicon work and I think we're good to go.

sleroux commented:
Also just found a small UI bug. When tapping on the URL bar then going in and out of the tab tray, the share button appears slightly below the input field on the right side..

https://www.dropbox.com/s/w2ja0n8ru2u11wa/Screenshot%202016-06-28%2010.30.22.png?dl=0

tylerlacroix commented:
I'll look into that UI bug!

sleroux commented:
Forgot to mention, before merging this we'll need to wrap it in a feature flag so we can turn it off in case we don't ship this feature.

sleroux commented:
Just ran this with your recent changes to add the feature flag. When turning it OFF, the tabs button doesn't seem to be working on iPad.

tylerlacroix commented:
Sorry, it looks like when I refactored the tabsButton into the TabTrayButtonExtensions I removed the addTaget.  My bad, the tab button should work now.

sleroux commented on Client/Frontend/Strings.swift:
```diff
@@ -184,3 +184,8 @@ extension Strings {
+
+// Top Tabs
+extension Strings {
+    public static let TopTabsNewTabTitle = NSLocalizedString("TopTabs.Tiltle.NewTab", value: "New Tab", comment: "The title displayed in the tab when opening a new tab.")
```
I would keep this string where it was instead of creating a new localization value for it since it will appear as a new string for the localizers. Also theres a type in TopTabs.Tiltle.NewTab.

tylerlacroix replied:
Okay, so I should remove this and add the string to the TopTabsViewController file?

sleroux replied:
I assume this string was refactored out of an existing button since we already had the text 'New Tab' available. We just can't add any additional strings with a new value for NSLocalizedString.

tylerlacroix replied:
Okay, I can just use the string from AppMenuConfiguration.NewTabTitleString
